### ISSUE TITLE
Document repair-based tombstone removal

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#14130]
I had in https://github.com/scylladb/scylladb/issues/14051 a discussion with a user that suspects they have too many tombstones and would have liked to get rid of them. I told them we have a new feature called "Repair-based tombstone GC" which can remove all the tombstone from his table without waiting for gc_grace_seconds. I found a blog post about it https://www.scylladb.com/2022/06/30/preventing-data-resurrection-with-repair-based-tombstone-garbage-collection/ and found it was added in Scylla 5.0 as experimental and GA since Scylla 5.2 - but wasn't able to find proper documentation for this feature. Do we have such a document?

A search in our documentation site found https://docs.scylladb.com/stable/kb/tombstones-flush.html but this document seems out of date - it asks you to manually manipulate gc_grace_second instead of using the new tombstone_gc configuration. It also mentions "large partition" for no reason (I think) - nothing it says is specifically about large partitions.

If we don't have a document about the new repair-based-gc we should write one. We should also link to it from relevant places (talking about gc_grace_second, etc.), and also remove old documents that suggest worse techniques to achieve this thing - like the above link.

CC @asias @annastuchlik

**Comment by asias** [comment]
We have doc here:  https://docs.scylladb.com/stable/cql/ddl.html#tombstones-gc-options

**Comment by nyh** [comment]
Thanks, but:
1. The explanation written there suggest that "tombstone_gc=repair" mode is a way to **avoid gc** until a repair is done but does not explicitly say the opposite - that all tombstones are GCed **when** a repair is done. In fact, the lack of such a clear statement shook my confidence now... Will all the tombstones be immediately removed during repair, or not? Maybe gc_grace_period still matters in that mode? I thought it didn't, but now I'm not sure :-(
2. This is a small part of a HUGE document, and is not findable - I didn't find it when I searched the keywords I thought of in either Google or our own documentation search engine.
3. We have a special document on how to get rid of tombstones, and this is what the user (in https://github.com/scylladb/scylladb/issues/14051) wanted to do, but instead of explaining the best way, it explains some old way.

**Comment by nyh** [comment]
I think I wasn't clear enough perhaps:

The document I am asking for would explain how to get rid of old tombstones - not how to avoid getting rid of old tombstones. If `tombstone_gc=repair` plus a repair is a good and safe way to do it, it should be explained explicitly, not inferred from some document that explains how to **avoid** dropping old tombstones.

**Comment by asias** [comment]
> Thanks, but:
> 
>     1. The explanation written there suggest that "tombstone_gc=repair" mode is a way to **avoid gc** until a repair is done but does not explicitly say the opposite - that all tombstones are GCed **when** a repair is done. In fact, the lack of such a clear statement shook my confidence now... Will all the tombstones be immediately removed during repair, or not? Maybe gc_grace_period still matters in that mode? I thought it didn't, but now I'm not sure :-(

No. The tombstones will not be removed immediately during or after repair. The gc_grace_seconds does not matter in tombstone_gc=repair any more. After repair, compaction can remove the tombstones. But repair does not guarantee compaction will do the removal immediately. 

> 
>     2. This is a small part of a HUGE document, and is not findable - I didn't find it when I searched the keywords I thought of in either Google or our own documentation search engine.

Yes, doc team is making it more searchable.  

> 
>     3. We have a special document on how to get rid of tombstones, and this is what the user (in [Scanning a table using consistent read fails, but succeeds on other isomorphic tables #14051](https://github.com/scylladb/scylladb/issues/14051)) wanted to do, but instead of explaining the best way, it explains some old way.

**Comment by nyh** [comment]
> No. The tombstones will not be removed immediately during or after repair. The gc_grace_seconds does not matter in tombstone_gc=repair any more. After repair, compaction can remove the tombstones. But repair does not guarantee compaction will do the removal immediately.

All these steps should be documented in on clear guide. I (and apparently, many users, see also a new issue #14484) want a clearly documented way to get rid of massive amounts of tombstones that appear after doing massive deletions, and want a document that explains clearly the steps needed to get rid of old tombstones. If these steps involve setting `tombstone_gc` , doing a repair, and then doing a compaction (what kind of compaction?), then this is exactly what this document should explain.

As I noted above, we have "almost" the right document, in https://opensource.docs.scylladb.com/stable/kb/tombstones-flush.html - but it's not up-to-date with the new `tombstone_gc` feature and it contains some details which seem fishy to me - why are "large partitions" are mentioned? With tombstone_gc=repair, do we really need **major** compaction? Why doesn't the document mention the fact that there are multiple nodes, so what needs to be run where?

**Comment by annastuchlik** [comment]
@nyh You're right that it's not searchable and that more documentation is needed to help the users. I'm going to work on it soon. I'll include more details provided in this PR by @asias .

> As I noted above, we have "almost" the right document, in https://opensource.docs.scylladb.com/stable/kb/tombstones-flush.html - but it's not up-to-date with the new tombstone_gc feature and it contains some details which seem fishy to me - why are "large partitions" are mentioned? With tombstone_gc=repair, do we really need major compaction? Why doesn't the document mention the fact that there are multiple nodes, so what needs to be run where?

@asias Could you please elaborate on the above so that I have complete information?

**Comment by raphaelsc** [comment]
As @nyh noted, we're missing documentation that explains propagation_delay setting in the repair tombstone gc mode: https://opensource.docs.scylladb.com/stable/cql/ddl.html#tombstones-gc-options

/cc @asias

**Comment by Dhruv-Garg79** [comment]
https://docs.scylladb.com/stable/cql/ddl.html#tombstones-gc-options

This documentation is still unclear for repair mode, can we please update it to include more details like what everyone has discussed here?
